homepage/tests/add_review_search_tests.py

```diff
@@ -0,0 +1,72 @@
+import pytest
+from homepage.models import Course
+from django.db.models import QuerySet
+from pytest_django.asserts import assertTemplateUsed
+
+
+# --------Backend testing-------- #
+@pytest.mark.django_db
+def test_type_get_courses_ordered_by_name():
+    # because an empty string is a substring of all names, all courses returned
+    courses = Course.get_courses_ordered_by_name("")
+
+    assert isinstance(courses, QuerySet)
+    assert all(isinstance(course, Course) for course in courses)
+
+
+@pytest.mark.parametrize("course_name_exist, expected_output", [
+    ("Resonance", ["Resonance in Runes and Signs"]),
+    ("Numer", ["Numerology"]),
+    ("in", ["Grammatica in Arithmancy", "Resonance in Runes and Signs", "UnFogging the Future"]),
+])
+@pytest.mark.django_db
+def test_name_exist_get_courses_ordered_by_name(course_name_exist, expected_output):
+    courses = Course.get_courses_ordered_by_name(course_name_exist)
+
+    assert all(course.name in expected_output for course in courses)
+
+
+@pytest.mark.parametrize("wrong_course_name", [
+    ("Introduction to Poetry"),
+    ("Introduction to Buddhism"),
+])
+@pytest.mark.django_db
+def test_wrong_name_get_courses_ordered_by_name(wrong_course_name):
+    courses = Course.get_courses_ordered_by_name(wrong_course_name)
+
+    assert not courses
+
+
+@pytest.mark.django_db
+def test_lexicography_order_get_courses_ordered_by_name():
+    courses = list(Course.get_courses_ordered_by_name(""))
+
+    for index in range(len(courses)-1):
+        assert courses[index].name < courses[index+1].name
+
+
+# --------Front End testing-------- #
+@pytest.mark.django_db
+def test_renders_add_review_search_template(client):
+    response = client.get('/add_review_search/')
+
+    assert response.status_code == 200
+    assertTemplateUsed(response, 'homepage/add_review_search.html')
+
+
+@pytest.mark.django_db
+def test_get_course_by_name_with_client(client):
+    response = client.get('/add_review_search/', {'course': 'Resonance'})
+    courses_not_found = [b'Grammatica in Arithmancy', b'Numerology', b'UnFogging the Future']
+
+    assert response.status_code == 200
+    assert b'Resonance' in response.content
+    assert all(course not in response.content for course in courses_not_found)

```

Comment:
Too many asserts IMO for a single function.
I would split this into 2 methods - 1 method for found courses and 1 method for not found courses.

Reply:
@LeonRabinovich Please take a closer look. 
When the user searches for a course, the expected result is that single course.
Therefore we must check that the rest of the courses are absent in the response.
BTW you can see the function for not found courses below.

Reply:
Not sure what you mean, i meant for you to simply split the same method you already have into 2 methods.

(1)
**def test_get_course_by_name_with_client(client):**
    response = client.get('/add_review_search/', {'course': 'Resonance'})

    assert response.status_code == 200
    assert b'Resonance' in response.content

(2)
**def test_courses_not_found_with_client(client):**
    response = client.get('/add_review_search/', {'course': 'Resonance'})
    courses_not_found = [b'Grammatica in Arithmancy', b'Numerology', b'UnFogging the Future']

    assert all(course not in response.content for course in courses_not_found)

Reply:
@LeonRabinovich I understand very well what you said in the first comment.
It makes no sense to split it into 2 tests as these checks are logically linked.
About the "Too many asserts", I invite you to take a look at the last lecture:
![image](https://user-images.githubusercontent.com/57869309/117545816-17d9ca80-b030-11eb-8ec3-3f5991eb1e68.png)
The code in the presentation has the same amount of asserts or even more, so it is completely fine.

Reply:
@LeonRabinovich **The example you mentioned is different, as you were checking two aspects in the same test.**
As I said before, if you look again at the code, **there is already a separate test for not found courses**.
I explained in the first comment why these checks must be in the same test. 
**You've said you didn't understand the explanation, but you insist on the change**. I find that weird. 
Please read the first comment and tell me what you didn't understand so I can explain.

Reply:
@LeonRabinovich, I agree with @oriAdler IMO those asserts are linked, and test the same results from different aspects - that the course was found, and only that course. Therefore, they should be together in the same test.

Reply:
@ellaml @oriAdler 
Ok, i approved the review